kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -387,48 +645,77 @@ string KvbAppFilter::readBlockRangeHash(BlockId block_id_start, BlockId block_id
-  if (not opt) {
+string KvbAppFilter::readEventGroupRangeHash(EventGroupId event_group_id_start) {
+  uint64_t public_start = getValueFromLatestTable(kPublicEgIdKeyOldest);
+  uint64_t public_end = getValueFromLatestTable(kPublicEgIdKeyNewest);
+  uint64_t private_start = getValueFromLatestTable(client_id_ + "_oldest");
+  uint64_t private_end = getValueFromLatestTable(client_id_ + "_newest");
+  if (!public_start && !private_start) {
     std::stringstream msg;
-    msg << "An event group for trid \"" << client_id_ << "\" cannot be found";
+    msg << "Event groups do not exist for client: " << client_id_ << " yet.";
+    LOG_ERROR(logger_, msg.str());
     throw std::runtime_error(msg.str());
+  } else if (!public_start) {
+    ConcordAssertEQ(public_start, public_end);
+  } else if (!private_start) {
+    ConcordAssertEQ(private_start, private_end);
   }
-  auto val = std::get_if<concord::kvbc::categorization::VersionedValue>(&(*opt));
-  if (not val) {
-    std::stringstream msg;
-    msg << "Couldn't convert stored event group id to versioned value";
-    throw std::runtime_error(msg.str());
+  uint64_t event_group_id_end = private_end + public_end;
+  if (event_group_id_start == 0) {
+    throw InvalidEventGroupId(event_group_id_start);
   }
-  auto last_trid_eg_id = concordUtils::fromBigEndianBuffer<uint64_t>(val->data.data());
-
-  if (event_group_id_start > event_group_id_end || event_group_id_end > last_trid_eg_id) {
+  if (event_group_id_start > event_group_id_end) {
     throw InvalidEventGroupRange(event_group_id_start, event_group_id_end);
   }
-
-  EventGroupId event_group_id(event_group_id_start);
-
-  LOG_DEBUG(logger_, "readEventGroupRangeHash event_group_id " << event_group_id << " to " << event_group_id_end);
-
   string concatenated_update_hashes;
-  concatenated_update_hashes.reserve((1 + event_group_id_end - event_group_id) * kExpectedSHA256HashLengthInBytes);
-  for (; event_group_id <= event_group_id_end; ++event_group_id) {
+  // we might reserve more than we need because we can have duplicate entries b/w [event_group_id_start,
+  // event_group_id_end]
+  concatenated_update_hashes.reserve((1 + event_group_id_end - event_group_id_start) *
+                                     kExpectedSHA256HashLengthInBytes);
+
+  // update the offsets if we now have corresponding public/private event groups in storage
+  if (eg_hash_state_->public_offset == 0) eg_hash_state_->public_offset = public_start;
+  if (eg_hash_state_->private_offset == 0) eg_hash_state_->private_offset = private_start;
+
+  // populate and read global event group ids from eg_data_state_->event_group_id_batch in batches of size kBatchSize.
+  // For every global event group id received, lookup the data table to fetch the event group, filter and calculate the
+  // concatenate hash from the filtered event group
+  while (eg_hash_state_->curr_trid_event_group_id < event_group_id_end) {
```

Comment:
I wonder if we can refactor most of this functionality (beginning plus while-loop) given that it is used in 3 functions. No need to do this now but this would be a good follow-up in order to keep code-duplication low.

Reply:
Yes, will update in a follow-up PR.